### ISSUE TITLE
setitimer effect and SIGALRM delivery are not sequentially consistent

### ISSUE THREAD

**Issue by xilun** [Microsoft/BashOnWindows#474]
Hi,
Here is an strace I managed to got (strace is still unrealiable in regard with signals on build 14352, so it was not easy, I had to call strace in a loop on a somehow misbehaving process...)

    Process 1941 attached
    18:44:10.788989 rt_sigaction(SIGALRM, {SIG_IGN, [ALRM], SA_RESTORER|SA_RESTART, 0x7fff9b066cb0}, {0x403080, [], SA_RESTORER, 0x7fff9b066cb0}, 8) = 0
    18:44:10.790302 recvfrom(4, "priseServices.Thunk.dll\r\n       "..., 16384, MSG_DONTWAIT, NULL, NULL) = 16384
    18:44:10.791385 rt_sigaction(SIGALRM, {0x403080, [], SA_RESTORER, 0x7fff9b066cb0}, NULL, 8) = 0
    18:44:10.792633 setitimer(ITIMER_REAL, {it_interval={0, 0}, it_value={0, 50000}}, NULL) = 0
    18:44:10.793382 --- SIGALRM {si_signo=SIGALRM, si_code=SI_USER, si_pid=0, si_uid=0} ---
    strace: ptrace(PTRACE_SYSCALL,pid:1941,sig:14): Invalid argument
    Process 1941 detached

The SIGALRM shown here can come from a previous call to setitimer not traced here. This trace shows that the effects of setitimer is not sequentially consistent with the delivery of the SIGALRM signals it triggers, otherwise SIGALRM would not have been delivered less than 1ms after a successful call to setitimer setting up an alarm for 50ms later...

**Comment by stehufntdev** [comment]
Thanks for reporting the issue, can you share out the steps you took so we can try to repro locally?  I looked through the itimer code and it should be canceling out any previously set timers.

**Comment by xilun** [comment]
Well I got to that through rather convoluted means by trying to work-around the fact that under Posix we are not supposed to switch std fd to non-blocking because that can also apply to other processes having the same inherited redirections (pb described here: https://cr.yp.to/unix/nonblock.html )

So I first gave a try to the "Interval timers" solution and the result is in [this branch of my tool](https://github.com/xilun/cbwin/tree/buggy_proto_suspend_resume). In particular [this commit](https://github.com/xilun/cbwin/commit/b5e70173f1fc2919fec88b59c70b551e73cb87ae) shows the code involved. (At one point disarm_alarm() was simpler, without the little sigprocmask() dance, but the end result was the same: sometimes the process consume 1 whole CPU even when it has nothing to do.)

In the end I could not go anywhere, and with strace quitting at the first signal it was quite hellish to debug, so I just switched to using threads. But before that, when trying to debug, I got that weird strace (before switching to threads, so this was in a completely mono-thread process).

I'm not even sure SIGALRM *deliveries* are supposed to be sequentially consistent with setitimer() calls, neither I have validated that they indeed are under Linux, but it really feels they should in order to be usable reliably. So I think two things can happen here: either your signal delivery mechanism is somehow asynchronous in regard to the completion of setitimer() syscalls (a SIGALRM triggered before the last setitimer() call could still be in flight and with no provision to cancel it, maybe?), or there is always the possibility that strace mixed up things or that I got a nasty time adjustment precisely at this point by vmware or whatever. Note that I also call setitimer() from the handler even if it is officially not Async-signal-safe, but I believe this is OK the way I do it (the handler is *not* SA_NODEFER, and setitimer() is otherwise only called at points where the handler can not execute.)

For now I don't have time to work on a more minimal repro (esp. given I worked-around the pb completely by using a different design), so if you are still interested I can give exact instructions for a repro with that branch of my tool (the part that runs under WSL is ~1000 lines of C)

**Comment by stehufntdev** [comment]
Thanks for the summary, sounds like this is not an easy repro.

Do you have a strace from the start of the program running?  The strace above appears to be an attach so it only collected data after the program was running.  I wanted to see if there was a previous call to setitimer that may help explain what happened so we can try to create a repro locally.

**Comment by xilun** [comment]
I don't have the strace from program start (because strace for now crashes when any signal happens on public build 14352) and I managed to got this one by running strace in a while true loop. But I'm pretty sure the process was experiencing a SIGLARM storm, so strace crashed all the time, so I only managed to capture short traces and this one was weird. There were definitively calls to setitimer() before, because there was a call to setitimer() before each write to stdout, and tons of data to forward with tons of writes.
